gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+        self._sc.unlock()
+
+
+class PollHandler(object):
+
+    def __init__(self, sc, qu, eh, batch=-1):
+        self._sc = sc
+        self._eh = eh
+        self._cache = EventCache(sc)
+        self._pollq = qu
+        self._procidx = 0
+        self._procpending = 0
+        self._batch = 10 if batch == -1 else batch
+
+    def add(self, event):
+        LOG.debug(_("Add event %s to the pollq" % (event.identify())))
+        self._pollq.put(event)
+
+    def rem(self, event):
+        LOG.debug(_("Remove event %s from pollq" % (event.identify())))
+        LOG.debug(
+            _("Removing all poll events with key %s" % (event.identify())))
+        remevs = []
+        cache = self._cache.copy()
+        for el in cache:
+            if el.key == event.key:
+                LOG.debug(_(
+                    "Event %s key matched event %s key - "
+                    "removing event %s from pollq"
+                    % (el.identify(), event.identify(), el.identify())))
+                remevs.append(el)
+        self._cache.rem_multi(remevs)
+
+    def _get(self):
+        try:
+            return self._pollq.get(timeout=0.1)
+        except Empty:
+            return None
+
+    def fill(self):
+        LOG.debug(_("Fill events from multi processing Q to internal cache"))
+        # Get some events from queue into cache
+        for i in range(0, 10):
+            ev = self._get()
+            if ev:
+                LOG.debug(_(
+                    "Got new event %s from multi processing Q"
+                    % (ev.identify())))
+                self._cache.add(ev)
+
+    def peek(self, idx, count):
+        LOG.debug(_("Peek poll events from index:%d count:%d" % (idx, count)))
+        cache = self._cache.copy()
+        qlen = len(cache)
+        LOG.debug(_("Number of elements in poll q - %d" % (qlen)))
+        pull = qlen if (idx + count) > qlen else count
+        return cache[idx:(idx + pull)], pull
+
+    def event_done(self, ev):
+        LOG.info(_("Poll event %s to be marked done !" % (ev.identify())))
+        self.rem(ev)
+
+    def _cancelled(self, ev):
+        LOG.info(_("Poll event %s cancelled" % (ev.identify())))
+        ev.poll_event = 'POLL_EVENT_CANCELLED'
+        self.event_done(ev)
+        self._sc.rpc_event(ev)
+
+    def _sched(self, ev):
+        LOG.debug(_("Schedule event %s" % (ev.identify())))
+        eh = self._eh.get(ev)
+        if isinstance(eh, core_periodic_task.PeriodicTasks):
+            if eh.check_timedout(ev):
+                LOG.info(_(
+                    "Event %s timed out -"
+                    "scheduling it to a worker" % (ev.identify())))
+                self._sc.rpc_event(ev)
+                return ev
+        else:
+            LOG.info(_(
+                "Event %s timed out -"
+                "scheduling it to a worker" % (ev.identify())))
+            self._sc.rpc_event(ev)
+            return ev
+        return None
+
+    def event(self, ev):
```

Comment:
Removed the "event" method and added logic in "_process_event" itself. Please check the latest commit.